Remove `structs.h` in favor of manual padding

Jan200101 opened the pull request:
Changes:
- delete `structs.h` ([hooray](https://www.youtube.com/watch?v=vVprmiy74ng))
- completely remove VPKFileEntry (its completely unused)
- convert all other instances of `OFFSET_STRUCT` to normal structs (this was automated and may be wrong)
  - Update `m_bZooming` type from `uint32_t` to `bool`

This is completely untested and may violently combust

Jan200101 commented on NorthstarDLL/engine/r2engine.h:
```diff
@@ -173,20 +173,29 @@ namespace R2
-		FIELD(0xF500, char m_UID[32])
-	};
-	// clang-format on
+		char _unk1[22]; // 0x0 ( Size: 22 )
```
correct

GeckoEidechse replied:
oh, `unk` is `unknown`, lol

GeckoEidechse replied:
Actually is `unk` a common abbreviation for `unknown` that I'm unaware of? Or is there a reason to use `unk` instead of the more self-explanatory `unknown`? ^^"

Jan200101 replied:
idk.
I picked it up while digging through decompiles and it stuck with me.
Used it here because it was nice and small

ASpoonPlaysGames replied:
> Actually is `unk` a common abbreviation for `unknown` that I'm unaware of? Or is there a reason to use `unk` instead of the more self-explanatory `unknown`? ^^"

I've seen both, I think it changes based on the decompiler used. `unk` is used over `unknown` because it's more concise, there isn't really much meaning behind saying that something is "unknown", the important part of the name is the offset that is appended

p0358 commented:
You really should, and I cannot underestimate this, add a static_assert for every member to ensure its offset (use `offsetof` macro).

Having such safety can allow you to drop `#pragma pack(push, 1)` and then drop some paddings to make the structs look less cluttered (and sometimes 1:1 to how it's laid out in Source SDK headers). Default packing is 8, but some things like two ints might be packed directly one after another.

Having such asserts has saved me many times in the past, and it can save you (and other contributors) too

Jan200101 commented:
appears to run the game just fine

p0358 commented:
until it doesn't and someone spends hours trying to figure out why

when I was adding those to my project I found a few fields that were wrong, but it didn't immediately obviously break stuff at a glance; similar situation during refactors when you merge or split up some classes, or have some inheritance with vtables like for ConVars, how do you ensure they're at their actual correct offset then? just close your eyes and hope it's okay? xd

Jan200101 commented:
> until it doesn't and someone spends hours trying to figure out why

to be clear, I was talking about the current state as it is.
I'll add the offset checks regardless of if it works or not, because its saner to have.

F1F7Y commented on NorthstarDLL/engine/r2engine.h:
```diff
@@ -202,63 +211,56 @@ namespace R2
+
+		const char* m_pMapName; // 0x60 ( Size: 8 )
+		int m_nMapVersion; // 0x68 ( Size: 4 )
+	} CGlobalVars;
```
`m_pMapName` and `m_nMapVersion` are in `CGlobalVars` ( second link ). `m_nTickCount` is missing because i didn't confirm its location. Regarding classes: they are the same as structs but members are private by default. If we want to mimic source we should mimic classes/structs as much as we can even though they're the same in cpp.

Jan200101 replied:
> Regarding classes: they are the same as structs but members are private by default.

I am still not a fan of it.
structs are far better for data, which most of it is and the few function pointers we need to deal with can be handled elsewhere.

I'll add it regardless

Jan200101 commented:
- added offset checks
- converted into class
- removed packing (except for the player class because it generates some padding somewhere and I cannot be arsed to find it)
- added some Primedev members

GeckoEidechse commented on NorthstarDLL/server/r2server.h:
```diff
@@ -9,102 +9,158 @@ namespace R2
+		Vector3 m_vecAbsOrigin; // 0x100490 ( Size: 12 )
+		char _unk57[7656]; // 0x10049c ( Size: 7656 )
+		Vector3 m_upDir; // 0x102284 ( Size: 12 )
+	} CBasePlayer;
```
I assume this was implemented @Jan200101? ^^

Jan200101 replied:
No.
Prime splits it up into multiple classes and then glues it together using inheritance and I am not to keen to just copy paste all over it over.

GeckoEidechse replied:
Gotcha. So I guess this conversation can be resolved then as not planned (for this PR)? ^^

Jan200101 replied:
email replies apparently don't work for reviews:
---
yeah

GeckoEidechse commented on NorthstarDLL/server/r2server.h:
```diff
@@ -9,102 +9,251 @@ namespace R2
-	// clang-format on
+#pragma pack(pop)
+
+	static_assert(offsetof(CBasePlayer, m_nPlayerIndex) == 0x58);
```
Definitely good thing to have the static asserts in there as well to ensure the compiler doesn't mess stuff up :D

GeckoEidechse reviewed (commented):
(not giving approve as this is more of an opinion than code review)

Looking at the changes, I have to say I kinda prefer the packed struct over the existing macro mess.

The original code calls the macro to be nicer to update but tbh, simply replacing an `_unk` with the newly reversed variable seems nicer to me than adding another field. Especially the more we reverse of the struct.

Jan200101 commented:
Yeah

13 Dec 2023 15:08:44 GeckoEidechse ***@***.***>:

>
> ***@***.**** commented on this pull request.
>
> ----------------------------------------
>
> In 
> NorthstarDLL/server/r2server.h[https://github.com/R2Northstar/NorthstarLauncher/pull/613#discussion_r1425407778]:
>
>> +     char m_title[32]; // 0x2848 ( Size: 32 )
> +       char _unk51[252]; // 0x2868 ( Size: 252 )
> +       bool m_useCredit; // 0x2964 ( Size: 1 )
> +       char _unk52[51]; // 0x2965 ( Size: 51 )
> +       float m_smartAmmoPreviousHighestLockOnMeFractionValue; // 
> 0x2998 ( Size: 4 )
> +       char _unk53[1292]; // 0x299c ( Size: 1292 )
> +       int32_t m_pilotClassIndex; // 0x2ea8 ( Size: 4 )
> +       char _unk54[1456]; // 0x2eac ( Size: 1456 )
> +       int32_t m_playerScriptNetDataGlobal; // 0x345c ( Size: 4 )
> +       char _unk55[5352]; // 0x3460 ( Size: 5352 )
> +       int32_t m_selectedOffhand; // 0x4948 ( Size: 4 )
> +       char _unk56[1030980]; // 0x494c ( Size: 1030980 )
> +       Vector3 m_vecAbsOrigin; // 0x100490 ( Size: 12 )
> +       char _unk57[7656]; // 0x10049c ( Size: 7656 )
> +       Vector3 m_upDir; // 0x102284 ( Size: 12 )
> +   } CBasePlayer;
>
> Gotcha. So I guess this conversation can be resolved then as not 
> planned (for this PR)? ^^
>
> —
> Reply to this email directly, view it on 
> GitHub[https://github.com/R2Northstar/NorthstarLauncher/pull/613#discussion_r1425407778], 
> or 
> unsubscribe[https://github.com/notifications/unsubscribe-auth/ADTAVLLIJTVFG2SLL6HOK3DYJGZGZAVCNFSM6AAAAABAF3TRZ6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTONZZG4ZDIOBZHA].
> You are receiving this because you were mentioned.[Tracking 
> image][https://github.com/notifications/beacon/ADTAVLMFO377DSZ5KLRU4ZTYJGZGZA5CNFSM6AAAAABAF3TRZ6WGG33NNVSW45C7OR4XAZNRKB2WY3CSMVYXKZLTORJGK5TJMV32UY3PNVWWK3TUL5UWJTTKCRZGE.gif]
>

p0358 commented:
I don't really have any further objections. Just one suggestion I thought of: if you named the _unks with its offset afterwards instead of an index number (like `_unk_0x2b8` instead of `_unk4`), it'd make it easier to later insert a new reversed struct member, without having to update all the changed ordered index numbers, making up for smaller diffs and better git blames afterwards

Jan200101 commented:
good point, will add

ASpoonPlaysGames commented:
Just to note this PR also solves like 100+ warnings when compiling with `/W4` (MSVC)

ASpoonPlaysGames reviewed (approved):
Code looks good to me, didn't validate every offset but all of the ones i looked at seem to match up with the old stuff so i'm going to trust the script that was used to generate them

GeckoEidechse commented:
(@Mauler125 said LGTM in discussion about it on Discord)

ASpoonPlaysGames commented:
I hate github web merge resolving, it just replaced all the tabs with spaces :(

ASpoonPlaysGames reviewed (approved):
Played a couple of games with the PR, seems to not violently combust?

Jan200101 commented:
poeggers

GeckoEidechse commented:
Merging based on reviews
(really hoping the offsets are correct ^^)
